src/transformers/modeling_utils.py

```diff
@@ -2934,12 +2938,31 @@ def _fix_key(key):
         missing_keys = list(set(expected_keys) - set(loaded_keys))
         unexpected_keys = list(set(loaded_keys) - set(expected_keys))
 
-        # Some tensors maybe have been already filled by another key (tied weights).
-        # TODO: Sylvain -> make this work even on meta device.
-        # existing_ptrs = {model_state_dict[k].data_ptr() for k in loaded_keys if k in model_state_dict}
-        # missing_keys = [
-        #     k for k in missing_keys if k in model_state_dict and model_state_dict[k].data_ptr() not in existing_ptrs
-        # ]
+        def _tensor_hash(tensor):
+            # This is better than `tensor.data_ptr()`
+            # Since A = torch.zeros((10, 10))
+            # B = A[2, :]
+            # Then A.data_ptr() != B.data_ptr()
+            # But actually the storage is still shared
+            try:
+                ptr = tensor.untyped_storage().data_ptr()
+            except AttributeError:
+                # Fallback for torch==1.10
+                try:
+                    ptr = tensor.storage().data_ptr()
+                except NotImplementedError:
+                    # Fallback for meta storage like in 2.0
+                    ptr = 0
+            return (ptr, tensor.device)
+
+        existing_ptrs = {
+            _tensor_hash(model_state_dict[k])
+            for k in loaded_keys
+            if k in model_state_dict and model_state_dict[k].device != torch.device("meta")
+        }
```

Comment:
Here the check is not right as it does not find the tied weights when the model is on the meta device (which is going to be the default ultimately to load without using RAM). The goal is to detect tied parameters in the model in any case, so we can rely on the hash for the model weights (there are no shared slices in Transformers models) for this test.

Reply:
I swapped for accelerate here.